src/App.tsx

```diff
@@ -1,26 +1,361 @@
+      case Filter.Completed:
+        return todosList.filter(item => item.completed);
+      default:
+        return todosList;
+    }
+  }, []);
+
+  const visibleTodos = useMemo(
+    () => filterTodos(todos, filter),
+    [todos, filter, filterTodos],
+  );
+
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+  };
+
+  const handleClick = (
+    event: React.MouseEvent<HTMLAnchorElement>,
+    filterBy: Filter,
+  ) => {
+    event.preventDefault();
+    setFilter(filterBy);
+  };
+
+  const complateTodo = (todo: Todo) => {
+    setDeletingTodoId([todo.id]);
+
+    const complateTodoFromServer = async () => {
+      try {
+        const response: Todo = await updateTodo({
+          ...todo,
+          completed: !todo.completed,
+        });
+
+        if (!response) {
+          throw new Error('Failed to update todo');
+        }
+
+        setTodos(prevTodos =>
+          prevTodos.map(item =>
+            item.id === response.id
+              ? { ...response, completed: response.completed }
+              : item,
+          ),
+        );
+      } catch {
+        showError(Erors.UpdateError);
+      } finally {
+        setDeletingTodoId(null);
+      }
+    };
+
+    complateTodoFromServer();
+  };
+
+  const complateAllTodo = () => {
+    const isEveryCompleted = todos.every(item => item.completed);
+
+    const toggleAllOnServer = async () => {
+      const todosToUpdate = todos.filter(
+        todo => todo.completed === isEveryCompleted,
+      );
+
+      if (todosToUpdate.length === 0) {
+        return;
+      }
+
+      const updatingIds = todosToUpdate.map(todo => todo.id);
+
+      setDeletingTodoId(updatingIds);
+
+      try {
+        const requests = todosToUpdate.map(todo =>
+          updateTodo({
+            ...todo,
+            completed: !todo.completed,
+          }),
+        );
+
+        const results = await Promise.allSettled(requests);
+
+        setTodos(prev =>
+          prev.map(todo => {
+            const indexInChanged = updatingIds.indexOf(todo.id);
+
+            if (indexInChanged === -1) {
+              return todo;
+            }
+
+            const serverResult = results[indexInChanged];
+
+            if (serverResult.status === 'fulfilled') {
+              return serverResult.value as Todo;
+            }
+
+            return todo;
+          }),
+        );
+
+        if (results.some(todo => todo.status === 'rejected')) {
+          throw new Error();
+        }
+      } catch {
+        showError(Erors.UpdateError);
+      } finally {
+        setDeletingTodoId(null);
+      }
+    };
+
+    toggleAllOnServer();
+  };
+
+  const removeTodo = (id: number) => {
+    setDeletingTodoId([id]);
+    const removeTodoFromServer = async () => {
+      try {
+        const response = (await deleteTodo(id)) as number;
+
+        if (response !== 1) {
+          throw new Error();
```

Comment:
Do you catch this error somewhere? If not it will lead to page crash. Generally - do not throw errors again if you don't catch and handle them later